pallene/to_ir.lua

```diff
@@ -205,6 +205,46 @@ function ToIR:convert_stat(cmds, stat)
 
         table.insert(cmds, ir.Cmd.For(stat.loc, v, start, limit, step, ir.Cmd.Seq(body)))
 
+    elseif tag == "ast.Stat.ForIn" then
+        local decls = stat.decls
+        local exps = stat.exps
+
+
+        -- local iter, st, ctrl = ...rhs
+        local v_iter = ir.add_local(self.func, "$iter", exps[1]._type)
+        self:exp_to_assignment(cmds, v_iter, exps[1])
+        local v_state = ir.add_local(self.func, "$st", exps[2]._type)
+        self:exp_to_assignment(cmds, v_state, exps[2])
+        local v_ctrl = ir.add_local(self.func, "$ctrl", exps[3]._type)
+        self:exp_to_assignment(cmds, v_ctrl, exps[3])
+
+        -- body of the `while true ... end` loop.
+        local body = {}
+
+        -- local i, x = iter(st, ctrl)
+        local itertype = exps[1]._type
+        local v_decls = {}
+        for _, decl in ipairs(decls) do
+            local v = ir.add_local(self.func, decl.name, decl._type)
+            self.loc_id_of_decl[decl] = v
+            table.insert(v_decls, v)
+        end
+
+        local xs = { ir.Value.LocalVar(v_state), ir.Value.LocalVar(v_ctrl) }
+        local iter_f_id = self.fun_id_of_decl[exps[1].var._name.decl]
+        table.insert(body, ir.Cmd.CallStatic(exps[1].loc, itertype, v_decls, iter_f_id, xs))
+
+        -- if i == nil then break
+        local v_cond = ir.add_local(self.func, "$cond", types.T.Boolean())
+        local then_ = {}; table.insert(then_, ir.Cmd.Break())
+        table.insert(body, ir.Cmd.IsNil(stat.loc, v_cond, ir.Value.LocalVar(v_decls[1])))
+        table.insert(body, ir.Cmd.If(stat.loc, ir.Value.LocalVar(v_cond), ir.Cmd.Seq(then_), ir.Cmd.Seq({})))
```

Comment:
I think we can use `ir.Cmd.If(stat.loc, ir.Value.LocalVar(v_cond), ir.Cmd.Break(), ir.Cmd.Nop())`, without those Seqs.